src/array_api_extra/_apply.py

```diff
@@ -0,0 +1,285 @@
+    if shape is None:
+        shapes = [xp.broadcast_shapes(*(arg.shape for arg in args))]
+    elif isinstance(shape, tuple) and all(isinstance(s, int) for s in shape):
+        shapes = [shape]
+    else:
+        shapes = list(shape)
+        multi_output = True
+
+    if dtype is None:
+        dtypes = [xp.result_type(*args)] * len(shapes)
+    elif multi_output:
+        if not isinstance(dtype, Sequence):
+            msg = "Got sequence of shapes but only one dtype"
+            raise TypeError(msg)
+        dtypes = list(dtype)  # pyright: ignore[reportUnknownArgumentType]
+    else:
+        if isinstance(dtype, Sequence):
+            msg = "Got single shape but multiple dtypes"
+            raise TypeError(msg)
+        dtypes = [dtype]
+
+    if len(shapes) != len(dtypes):
+        msg = f"Got {len(shapes)} shapes and {len(dtypes)} dtypes"
+        raise ValueError(msg)
+    if len(shapes) == 0:
+        msg = "func must return one or more output arrays"
+        raise ValueError(msg)
+    del shape
+    del dtype
+
+    # Backend-specific branches
+    if is_dask_namespace(xp):
+        import dask  # type: ignore[import-not-found]  # pylint: disable=import-outside-toplevel,import-error  # pyright: ignore[reportMissingImports]
+
+        metas = [arg._meta for arg in args if hasattr(arg, "_meta")]  # pylint: disable=protected-access
+        meta_xp = array_namespace(*metas)
+
+        wrapped = dask.delayed(_npfunc_wrapper(func, multi_output, meta_xp), pure=True)
+        # This finalizes each arg, which is the same as arg.rechunk(-1)
+        # Please read docstring above for why we're not using
+        # dask.array.map_blocks or dask.array.blockwise!
+        delayed_out = wrapped(*args, **kwargs)
+
+        out = tuple(
+            xp.from_delayed(delayed_out[i], shape=shape, dtype=dtype, meta=metas[0])
+            for i, (shape, dtype) in enumerate(zip(shapes, dtypes, strict=True))
+        )
+
+    elif is_jax_namespace(xp):
+        # If we're inside jax.jit, we can't eagerly convert
+        # the JAX tracer objects to numpy.
+        # Instead, we delay calling wrapped, which will receive
+        # as arguments and will return JAX eager arrays.
+
+        import jax  # type: ignore[import-not-found]  # pylint: disable=import-outside-toplevel,import-error  # pyright: ignore[reportMissingImports]
+
+        wrapped = _npfunc_wrapper(func, multi_output, xp)
+        out = cast(
+            tuple[Array, ...],
+            jax.pure_callback(
+                wrapped,
+                tuple(
+                    jax.ShapeDtypeStruct(s, dt)  # pyright: ignore[reportUnknownArgumentType]
+                    for s, dt in zip(shapes, dtypes, strict=True)
+                ),
+                *args,
+                **kwargs,
+            ),
+        )
+
+    else:
+        # Eager backends
+        wrapped = _npfunc_wrapper(func, multi_output, xp)
+        out = wrapped(*args, **kwargs)
+
+        # Output validation
```

Comment:
I hope this doesn't introduce too much overhead for NumPy arrays, there's a lot of checks here compared to a regular function call. May make sense to measure once you have a branch that uses this on some scipy functionality.

Reply:
I'm happy to remove this